.github/workflows/ci.yml

```diff
@@ -109,10 +109,109 @@ jobs:
           COMMIT_MESSAGE: ${{ github.event.head_commit.message }}
           COMMIT_AUTHOR: ${{ github.event.head_commit.author.username }}
 
+  build-base:
+    name: base (${{ matrix.image }})
+    runs-on: ubuntu-latest
+    needs: [shellcheck, test, check]
+    if: github.event_name == 'push'
+    strategy:
+      fail-fast: false
+      matrix:
+        image:
+          - ubuntu
+          - centos
+          - emscripten
+    outputs:
+      images: ${{ steps.build-docker-image.outputs.images && fromJSON(steps.build-docker-image.outputs.images)  }}
+      coverage-artifact: ${{ steps.cov.outputs.artifact-name }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: ./.github/actions/setup-rust
+
+      - name: Set up Docker Buildx
+        if: runner.os == 'Linux'
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build xtask
+        run: cargo build -p xtask
+
+      - name: Prepare Meta
+        id: prepare-meta
+        timeout-minutes: 60
+        run: cargo xtask ci-job prepare-meta "${IMAGE}"
+        env:
+          IMAGE: ${{ matrix.image }}
+        shell: bash
+
+      - name: LLVM instrument coverage
+        uses: ./.github/actions/cargo-llvm-cov
+        with:
+          name: cross-${{matrix.image}}
+
+      - name: Install cross
+        if: matrix.deploy
+        run: cargo install --path . --force --debug
+
+      - name: Docker Meta
+        id: docker-meta
+        uses: docker/metadata-action@v4
+        with:
+          images: |
+            name=${{ steps.prepare-meta.outputs.image }}
+          labels: |
+            ${{ fromJSON(steps.prepare-meta.outputs.labels) }}
+      - name: Build Docker image
+        id: build-docker-image
+        timeout-minutes: 60
+        run: cargo xtask build-docker-image -v "${IMAGE}"
+        env:
+          IMAGE: ${{ matrix.image }}
+          LABELS: ${{ steps.docker-meta.outputs.labels }}
+          LATEST: ${{ needs.check.outputs.is-latest || 'false' }}
+        shell: bash
+
+      - name: Save Docker Image
+        id: save-docker-image
+        run: docker save "${IMAGE}:${TAG}" -o "${IMAGE}-${TAG}.tar"
+        env:
+            IMAGE: ${{ matrix.image }}
+            TAG: ${{ env.GITHUB_REF_NAME }}
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.image }}-image-tarball
+          path: ${{ matrix.image }}-${{ env.GITHUB_REF_NAME }}.tar
+
+      - uses: ./.github/actions/cargo-install-upload-artifacts
+        if: matrix.deploy
+        with:
+          target: ${{ matrix.image }}
+          image: ${{ steps.build-docker-image.outputs.image }}
```

Comment:
this is not needed, always skipped here and is only used to deploy binary to release